cmd/process-agent/app/status.go

```diff
@@ -0,0 +1,225 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package app
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"text/template"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	"github.com/DataDog/datadog-agent/cmd/process-agent/api"
+	"github.com/DataDog/datadog-agent/pkg/api/util"
+	ddconfig "github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/process/config"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+var httpClient = util.GetClient(false)
+
+const (
+	statusTemplate = `
+==============================
+Process Agent ({{ .Core.AgentVersion }})
+==============================
+
+ Status date: {{ .Date }}
+ Process Agent Start: {{ .Expvars.Uptime }}
+ Pid: {{ .Expvars.Pid }}
+ Go Version: {{ .Core.GoVersion }}
+ Python Version: {{ .Core.PythonVersion }}
+ Build arch: {{ .Core.Arch }}
+ Log Level: {{ .Core.Config.LogLevel }}
+ Enabled Checks: {{ .Expvars.EnabledChecks }}
+
+=========
+Collector
+=========
+
+  Last collection time: {{.Expvars.LastCollectTime}}
+  Docker socket: {{.Expvars.DockerSocket}}
+  Number of processes: {{.Expvars.ProcessCount}}
+  Number of containers: {{.Expvars.ContainerCount}}
+  Process Queue length: {{.Expvars.ProcessQueueSize}}
+  RTProcess Queue length: {{.Expvars.RTProcessQueueSize}}
+  Pod Queue length: {{.Expvars.PodQueueSize}}
+  Process Bytes enqueued: {{.Expvars.ProcessQueueBytes}}
+  RTProcess Bytes enqueued: {{.Expvars.RTProcessQueueBytes}}
+  Pod Bytes enqueued: {{.Expvars.PodQueueBytes}}
+
+`
+	notRunning = `
+=============
+Process Agent
+=============
+
+  The Process Agent is not running
+
+`
+)
+
+type coreStatus struct {
+	AgentVersion  string `json:"version"`
+	GoVersion     string `json:"go_version"`
+	PythonVersion string `json:"python_version"`
+	Arch          string `json:"build_arch"`
+	Config        struct {
+		LogLevel string `json:"log_level"`
+	} `json:"config"`
+	Metadata struct {
+		Meta struct {
+			Hostname string `json:"Hostname"`
+		} `json:"meta"`
+	} `json:"metadata"`
+	Endpoints map[string][]string `json:"endpointsInfos"`
+}
+
+type infoVersion struct {
+	Version   string
+	GitCommit string
+	GitBranch string
+	BuildDate string
+	GoVersion string
+}
+
+type processExpvars struct {
+	Pid                 int                    `json:"pid"`
+	Uptime              int                    `json:"uptime"`
+	MemStats            struct{ Alloc uint64 } `json:"memstats"`
```

Comment:
any reason why this is not a separately defined struct?

Reply:
Ah, the reason is because `Alloc` is a built-in expvar, and is nested within `MemStats`. This means that the easiest way to deserialize it is to define an anonymous struct that contains `Alloc`